Stop the capture map tools automatically (refs #4237)

SebDieBln opened the pull request:
Stops the capturing when leaving edit mode for the current layer or when changing the current layer. Avoids _ghost lines_ and other weird effects ([Redmine # 4237](https://hub.qgis.org/issues/4237)) when starting to digitize on one layer, but finishing it on another one (e.g. projection issues).

SebDieBln commented:
Just for reference I am linking to the - rejected - PR #1985, that was a different approach to this problem.

3nids commented:
I would be in favor of stopping digitizing when turning off editing but not sure about changing current layer.
I think this is an expected feature to be able to change current layer and come back to editing.
Wouldn't it be better to just allow editing to the same layer as the original one (to avoid finishing on another layer)?

Also, in both cases it would be nice to print messages to the user so he knows what's going on.

Opinions?

SebDieBln commented:
I agree that
> be able to change current layer and come back to editing

is a nice and probably expected feature.

However, the current implementation treats capturing as a global task, not a per layer one. The logical conclusion would be to make capturing a per layer task. So each (vector-)layer would have its own capturing state and only the one for the current layer would be shown.

Should I adapt the PR to implement this behaviour?

3nids commented:
An intermediate approach would be to save only one digitizing for a single layer. If you start digitizing on another layer, you prompt the user and tell him it will delete the first digitizing operation?

SebDieBln commented:
I will check which approach is easier to implement and modify the PR. Thanks for the suggestions!

SebDieBln commented:
Well, having multiple capturing processes in parallel is harder than I thought. So far it works for the _add feature_ tools and some of the feature moddifiying tools, too. But the _add circular string_ tools are giving me a headache and there are still some crashes now and then.

I will move this approach to a separate branch and go for the intermediate solution asking the user when the capturing would be interrupted.

Just give me a few more days.

SebDieBln commented:
OK, that is the simple solution now:
* When a new project is loaded/created, the digitizing tools are reset, thereby removing the dangling rubberbands.
* The behaviour when toggeling edit mode or switching the current layer is unchanged.

@3nids Please have a look at it.

My other approach of parallelizing things is now living in [its own branch](https://github.com/SebDieBln/QGIS/tree/MultiCapture).

3nids commented:
is `layerChanged` the most appropriate signal?
Adding layer should not stop capturing I would say.

I think that having the capturing active on a single layer was a good approach. Then you can stop capturing when current map tool layer is removed.

What do you think?

SebDieBln committed:
Reset digitizing tools when new project is loaded/created
(refs #4237)

SebDieBln commented:
Right, adding a layer should not stop capturing. I wanted to go the easy way and use a signal from the canvas. Glad you found it.
Since ```QgisApp``` is not reachable from the base class (it's in GUI), I added the handlers to the individual tools (they are in APP) and now only when loading/creating a project the tools are stopped.

> I think that having the capturing active on a single layer was a good approach.

Yes, but while working on these tools it turned out they pretty much do their own thing, very little is common, although the class hierarchy might suggest otherwise. E.g. I tried adding ESC-key functionality to the circular string tools for a different PR. But it only works for one of them, the other one does its own thing and is still immune to ESC, although implemented in the common base class.

So I would like to leave it at that for the moment and put some more work into the tools themselves. Even if this means the mentioned bug is not completely fixed yet. I therefore changed the title from _fixes_ to _refs_.

3nids commented:
There is only one map tool active at a time. Why not creating a slot in qgisapp to check if the current map tool is a capturing tool and if yes do stopCapturing?

SebDieBln commented:
> There is only one map tool active at a time.

Yes. But a capturing process is not aborted when a capturing tool is deactivated, see the rejected #1985 that wanted to implement this behaviour. So while at most one capturing tool can be active at a given time, each capturing tool can have its own running capturing process.

Three capturing processes running (add feature, split feature, split part) with no capturing tool is active:
![multiplecaptureprocesses](https://cloud.githubusercontent.com/assets/15147421/11984342/ac4e5eba-a9ba-11e5-91c7-74867e9c45f1.png)
.

SebDieBln commented:
@3nids I am not sure if you expect some modifications from me or if you are just busy. The latter is totally fine by all means, if it is the former, please let me know.

3nids commented:
I was waiting for a second advice.
Sorry for the delay, I will take care of this asap.

m-kuhn commented on src/app/qgsmaptooladdring.cpp:
```diff
@@ -29,6 +29,8 @@ QgsMapToolAddRing::QgsMapToolAddRing( QgsMapCanvas* canvas )
 {
   mToolName = tr( "Add ring" );
+  connect( QgisApp::instance(), SIGNAL( newProject() ), this, SLOT( stopCapturing() ) );
+  connect( QgisApp::instance(), SIGNAL( projectRead() ), this, SLOT( stopCapturing() ) );
```
What's you opinion about moving this to the `QgsMapToolCapture` constructor?

m-kuhn replied:
Hmmm... Would it maybe make more sense to connect the layer's `destroyed` signal instead?

SebDieBln replied:
@m-kuhn wrote:
> What's your opinion about moving this to the `QgsMapToolCapture` constructor?

@SebDieBln commented on 16 Dec 2015 
> Since `QgisApp` is not reachable from the base class (it's in GUI), I added the handlers to the individual tools (they are in APP).

Is there any workaround for this problem?

m-kuhn replied:
Nope, and more tools will be moved to gui.

Better go for signals that are available from layer/mapcanvas/maplayerregistry... like `QgsVectorLayer::destroyed()`

SebDieBln replied:
> Would it maybe make more sense to connect the layer's destroyed signal instead?

At the moment, a capturing process is not bound to a layer. This PR only targets the issue of orphaned rubberbands after loading a project/creating a new project.
Disallowing to start digitizing on layer A and finish it on layer B is a separate improvement.

m-kuhn replied:
But I guess removing the layer while capturing on it is the root problem. Doesn't the same apply when you remove a layer while capturing?

SebDieBln replied:
Removing the layer while in the middle of a capturing process is fine, capturing persists even without a layer. And that is the problem, because there is no `QgsVectorLayer` to connect to.

m-kuhn replied:
Sorry, I don't understand. Why should a capture process persist if the layer it is currently capturing features for (e.g. QgsMapToolAddFeature) is removed?

SebDieBln replied:
I agree it shouldn't persist, but it always did and it does at the moment. 
* Create layer A
* Start capturing
* Remove layer A
* Add layer B
* Continue and finish capturing
* Everything is fine

While I think binding the capture process to a layer is a good idea, it is more than I can do for 2.14.

m-kuhn commented on src/gui/qgsmaptoolcapture.h:
```diff
@@ -137,6 +132,12 @@ class GUI_EXPORT QgsMapToolCapture : public QgsMapToolAdvancedDigitizing
+    /**
+     * Stop capturing
+     */
+    void stopCapturing();
```
Please add a

    @note TODO QGIS 3: make protected

instead, this may break some things

SebDieBln replied:
`stopCapturing()` is already protected. The change is to make it a slot. Why would that break things?

m-kuhn replied:
Ah, I thought it was public. Sorry.

SebDieBln replied:
Sure, SIP bindings now got the update they deserved.

m-kuhn commented:
Concerning a plan to fix things properly in the long run.

When a maptoolcapture is activated:

 * it will remember which layer the operation happens on
 * connect the destroyed signal of this layer to deactivate
 * disconnect the signal in deactivate
 * when a click happens with this maptool on another layer ... deactivate()/activate() on new layer? continue on the originally active layer?

SebDieBln commented:
:+1: I am with you there 100%! In the long run.

m-kuhn commented:
Any comment on the layer change behavior?

SebDieBln commented:
`deactivate()` does not stop the capturing (the rejected #1985 wanted to implement that), so `stopCapturing()` / `startCapturing()` would be the way. Of course, `stopCapturing()` discards the work in progress. Maybe the user should be asked if that is what he wants to do?

Continuing on the originally active layer could be surprising for the user.

I'd vote for the the first alternative.

I also did some experiments implementing a separate parallel capture process for each layer. It still needs some work. What do you think of this approach?

m-kuhn commented:
That is, have multiple instances of a maptool alive in parallel?

m-kuhn commented:
@3nids would be good to have your feedback in here as well

SebDieBln commented:
> That is, have multiple instances of a maptool alive in parallel?
I was going for having multiple rubberbands inside a maptool, one for each layer, but the idea is the same, yes.

m-kuhn commented:
I wonder how this affects UX. 3 half-digitizied features on different layers may also be more confusing than good...

SebDieBln commented:
Probably best UX would be to ask the user and then `stopCapturing()` / `startCapturing()`.

m-kuhn commented:
Or re-select the currently active layer? I think it's much more common to continue the half-done job than to start  a new one. And in case he really wanted to start a new one he can cleanly abort the current one and change the layer again.

SebDieBln commented:
So using a maptool would change the active layer. I'd consider that a surprise to the user, and therefore I keep voting to ask the user to choose:
* Abort capturing on the other layer and start new capturing on the active layer.
* Continue capturing on the other layer.
* Cancel, e.g. don't do anything.

Could be improved by deactivating an active capturing maptool when changing the active layer and asking when activating the maptool, instead of keeping it active and asking when using the maptool.

m-kuhn commented:
The way I see visualize this before my mental eye:

 * Start adding a feature on layer A
 * Change maptool
 * Do something on layer B (still some rubberband from job A)
 * Reactivate the maptool add feature

In this scenario, I think it is much more common that I want to finalize the feature which I started before. So if in 90% switching the layer is the desired action and in only 10% another 2 clicks are required that sounds reasonable to me.

The main point against a question is, that it has to be a blocking and we should really only do this when there's no other way than a showing dialog.

SebDieBln commented:
You left out the _Activate layer B_ step before _Do something on layer B_. And you left out the corresponding _Activate layer A_ step. If the user correctly applies the _Activate layer A_ step, no dialog is ever shown. And therefore I still vote for asking the user if he left out this step, because I think this is a case where it is justified. :smile:

m-kuhn commented:
This just forces people to click more. Either on the legend entry or a message box. Where is the advantage? :)

SebDieBln commented:
The user is used to activate a layer before working with it. I don't see the need to save a click here.

Where do I see the advantage?
The capturing tools introduce [modes](https://en.wikipedia.org/wiki/Mode_(computer_interface)) to the interface, which give room for [mode errors](https://en.wikipedia.org/wiki/Mode_%28computer_interface%29#Mode_errors). Is there a current capturing process (and if yes, on what layer)? The user can't tell from the interface because the rubberband might be off-screen. The dialog informs the user about the mode.

Another solution that came to my mind just now was changing the icons for the maptools to be different whether the tool will _start a new capturing process_ or _continue an already begun capturing process_. So at least the fact that the user will continue a capturing process (which might include changing the active layer) is made clear in the interface.
This also covers the case where the user did not change the active layer but forgot he already started capturing. So I like this solution even better than the one with the dialog!

m-kuhn commented:
> The user is used to activate a layer before working with it.

Here we don't talk about "before", we talk about "while".

> Another solution that came to my mind just now was changing the icons for the maptools to be different whether the tool will start a new capturing process or continue an already begun capturing process.

An icon which is shown only very seldom won't be familiar to the user. I still prefer the "one operation at a time" , "no dialog" option. Let's wait what others (@3nids, @NathanW2 @nyalldawson ) think.

SebDieBln commented:
> Here we don't talk about "before", we talk about "while".

I disagree. "Do something on layer B" is "Work with layer B" and "Continue capturing on layer A" is "Work with layer A". If we talk "while", the user can't change the active layer from A to B in the first place.

> Let's wait what others (@3nids, @NathanW2 @nyalldawson ) think.

I am very interested to hear their opinions, too.
 
How about moving the discussion about the future behaviour of the capturing tools out of this PR to the mailing list?

My PR is just about a bugfix because of the rubberbands not being cleared when a new project is created/loaded. It does not introduce any serious change in the API, only one protected function upgraded to a protected slot. So whatever the decision will be, I assume it will work well with this bugfix.

3nids commented:
Hi, sorry for the silence.
My preferred approach would be to allow several digitizing process, one per layer all saved in the same instance of the map tool. 
I would only display one rubber band at a time, the one of the active layer.
I like the idea of the different icon when digitizing already started. Don't make it too different, maybe just a change of color.
I would avoid any asking to the user and always go for continuing digitizing. It's pretty quick and easy to cancel, and it should not be too frightening to the user.

m-kuhn commented:
I have just merged this manually.
It does make sense to recover leftover rubberbands by re-loading a project.

Anything else discussed in this pull request will be much appreciated. Thanks @SebDieBln and @3nids
